DOCS-4275 OAuth Provider Reference (#538)
mulesoft/docs-connectors: modules/ROOT/pages/oauth/oauth2-provider-documentation-reference.adoc

| Type a| Enumeration, one of:

** CONFIDENTIAL
** PUBLIC | The type of the client. Allowed values are PUBLIC: Clients incapable of maintaining the confidentiality of their credentials, or CONFIDENTIAL: Clients capable of maintaining the confidentiality of their credentials. |  PUBLIC | 
| Secret a| String |  |  | 
| Client Name a| String |  A friendly name for the client. |  | 
| Description a| String |  A brief description of the client. |  | 
| Principal a| String |  An optional principal to use when the ID can't be used with the security provider. |  | 
| Redirect Uris a| Array of String |  An expression that resolves to a list of redirect URIs used for when the client makes requests to the OAuth Provider. |  | 
| Authorized Grant Types a| Array of Enumeration, one of:

** AUTHORIZATION_CODE
** REFRESH_TOKEN
** TOKEN
** PASSWORD
** CLIENT_CREDENTIALS |  An expression that resolves to a list of the authorized grant types that the client can use to request a token.  |  | 
| Scopes a| Array of String |  An expression that resolves to a list of supported scopes by the client. If none are provided, the default scopes of the <<general-configuration>> are used. |  | 

> | Fail If Present a| Boolean |  Defines what to do if the a client with the same ID is already registered. If `true`, an error is raised. Otherwise, the client is updated.|  `false` | 

Reviewer: ```suggestion
| Fail If Present a| Boolean |  Defines what to do if a client with the same ID is already registered. If `true`, an error is raised. Otherwise, the client is updated.|  `false` | 
```

</oauth2-provider:clients>
----

Each registered client has an entry with the following information:

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Client Id a| String | A unique ID that will used to reference the client |  | x
| Principal a| String | An optional principal to use when the ID can't be used with the security
provider. For some security providers, the clientId can't be used for the client username. In those cases, the client's principal is used for authentication. |  | 
| Client Name a| String | The name of the client |  | 
| Description a| String | A short description for the client |  | 
| Secret a| String | A client secret used to authenticate the client. Only required if *Type* is CONFIDENTIAL. |  | 

> | Client Redirect Uris a| Array of String | List of registered redirectUris to use to respond to the client once the request has been processed.

Reviewer: ```suggestion
| Client Redirect Uris a| Array of String | List of registered redirect URIs to use to respond to the client once the request is processed.
```

> Most request allow the definition of new redirection URIs so these will not always be taken into account. See also <<newvaluehandling,New Value Handling>>. 

Reviewer: ```suggestion
Most requests allow the definition of new redirection URIs so these are not always taken into account. See also <<newvaluehandling,New Value Handling>>. 
```

`<oauth2-provider:client-scope value="USER"/>`|  | 
| Client Authorized Grant Types a| Array of Enumeration, one of:

** AUTHORIZATION_CODE
** REFRESH_TOKEN
** TOKEN
** PASSWORD
** CLIENT_CREDENTIALS | The grant types that define which OAuth flows this client will be able to be successfully execute to get a token. See also <<newvaluehandling,New Value Handling>>. |  | 

> | Client Scopes a| Array of String | The scopes that will match this client. If a request is received with this clientId but with different scopes, it will not be processed. See also <<newvaluehandling,New Value Handling>>. |  | 

Reviewer: ```suggestion
| Client Scopes a| Array of String | The scopes that will match this client. If a request is received with a matching client ID but with different scopes, it will not be processed. See also <<newvaluehandling,New Value Handling>>. |  | 
```

| Type a| Enumeration, one of:

** CONFIDENTIAL

> ** PUBLIC | The type of the client. Allowed values are PUBLIC  where clients incapable of maintaining the confidentiality of their credentials or CONFIDENTIAL where clients are capable of maintaining the confidentiality of their credentials. The client type defines if the client is able to maintain confidentiality for its credentials. | PUBLIC | 

Reviewer: ```suggestion
** PUBLIC | The client type defines if the client is able to maintain confidentiality for its credentials. Allowed values are `PUBLIC`, where clients do not maintain the confidentiality of their credentials, or `CONFIDENTIAL`, where clients maintain the confidentiality of their credentials.  | `PUBLIC` | 
```

|===

[[newvaluehandling]]

> For Client Redirect Uris, Client Authorized Grant Types, or Client Scopes, give each new value a new XML tag:

Reviewer: ```suggestion
For client redirect URIs, client authorized grant types, or client scopes, give each new value a new XML tag:
```

Reviewer: What happened to the XML tag for client scopes?

`<oauth2-provider:client-redirect-uri value="http://fake/redirect"/>`

`<oauth2-provider:client-authorized-grant-type value="AUTHORIZATION_CODE"/>`

[[ExpirationPolicy]]
=== Expiration Policy

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Max Idle Time a| Number | A scalar time value for the maximum amount of time a dynamic configuration instance should be allowed to be idle before it's considered eligible for expiration |  | 
| Time Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS | A time unit that qualifies the maxIdleTime attribute |  | 
|===

[[noRefreshToken]]
=== No Refresh Token

No refresh token is granted with every access token. As a consequence, 

> when a refresh token request comes, it always is rejected.

Reviewer: ```suggestion
when a refresh token request comes, it is always rejected.
```

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Token Generator Strategy a| TokenGeneratorStrategy |  |  | 
|===

[[singleRefreshToken]]
=== Single Refresh Token

For every new access token that is granted, a single refresh token is associated with it. That 
same refresh token can be used every time the access token is refreshed.

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Object Store a| Object Store | A reference to a globally defined object store or a definition of a private object store for storing generated refresh tokens. | A persistent object store created from the ObjectStoreManager with an entry TTL of 86400 SECONDS. | 
|===

[[multipleRefreshTokens]]
=== Multiple Refresh Tokens

A new refresh token is generated every time a refresh token request is executed. After that, the previous refresh token is invalidated.

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required

> | Object Store a| Object Store | A reference to a globally defined object store or a definition of a private object store, which stores the generated refresh tokens. | A persistent object store created from the ObjectStoreManager with an entry TTL of 86400 seconds. | 

Reviewer: ```suggestion
| Object Store a| Object Store | A reference to a globally defined object store, or a definition of a private object store that stores the generated refresh tokens. | A persistent object store created from the ObjectStoreManager with an entry TTL of 86400 seconds. | 
```

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Duration a| Number | The time to wait before resetting the rate limiter. That means that during time intervals of _duration_ length, every time a client validation fails, it will be added to the failure count. | 600 | 
| Duration Time Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  | SECONDS | 
| Maximum Failure Count a| Number | Maximum number of failures allowed within the period before preemptively rejecting requests. | 5 | 
|===

== OAuth Dance

Because the OAuth dance is done through HTTP, the OAuth2 Provider makes use of the xref:http/http-connector.adoc[Mule HTTP Connector].

As a consequence, apart from the definition of an OAuth2 provider configuration, the Mule application must also have an HTTP Listener configuration to be used by the provider.

Once configured, the provider works as follows:

Two HTTP endpoints are created for listening to Authentication Code and Token request as stated by the OAuth2 definition. Those work independently from the Mule application and respond via HTTP.

> Apart from that, the provider defines an operation: <<validateToken>> that can check if a token is authorized. That operation can be added anywhere in a flow to control it's execution. If the token is indeed authorized,

Reviewer: ```suggestion
The provider defines an operation: <<validateToken>> that can check if a token is authorized. That operation can be added anywhere in a flow to control its execution. If the token is indeed authorized,
```

> the flow will keep executing, setting token information in the payload, otherwise, a TOKEN_UNAUTHORIZED error will be raised. You need to add that operation in parts of the application that need token authorization.

Reviewer: ```suggestion
the flow continues executing, setting token information in the payload; otherwise, a `TOKEN_UNAUTHORIZED` error is raised. You need to add the operation to the parts of the application that require token authorization.
```

> Also, since token validation is almost always going to be used together with an HTTP Listener, in case it fails, the listener's response mechanism can handle that error and properly respond to the requester. Additional logic can be added handling that type of error.

Reviewer: ```suggestion
Since token validation is almost always used together with an HTTP Listener, in case it fails, the Listener's response mechanism can handle the error and properly respond to the requester. Additional logic can be added for handling that type of error.
```

Lastly, additional operations are provided to add or delete clients and to revoke tokens if needed.

== Security Providers

As seen in the <<general-configuration>>, two security providers should be defined in the app to be later referenced by the OAuth2 Configuration element.

> One way of doing this is using the Spring Framework, defining both security providers and then using the xref:spring/spring-module.adoc[Spring Module] to add the providers to the Mule Security Manager:

Reviewer: ```suggestion
One way of doing this is using the Spring Framework, defining both security providers, and then using the xref:spring/spring-module.adoc[Spring Module] to add the providers to the Mule Security Manager:
```
